src/Form/MetadataDisplayForm.php

```diff
@@ -254,38 +284,92 @@ public static function ajaxPreview($form, FormStateInterface $form_state) {
           'mode' => 'application/json',
         ],
       ];
-      sort($used_vars);
-      $json_keys = array_keys($jsondata);
-      $data_json = array_map(function($key) {
-        return 'data.' . $key;
-      }, $json_keys);
-      $unused_vars = array_diff($data_json,$used_vars);
-      sort($unused_vars);
-      $used_rows = array_map(function($used) {
-        return [$used];
-      }, $used_vars);
-      $unused_rows = array_map(function($unused) {
-        return [$unused];
-      }, $unused_vars);
-      $var_table = [
-        '#type' => 'table',
-        '#header' => [t('Used Variables')],
-        '#rows' => $used_rows,
-        '#empty' => t('No content has been found.'),
-      ];
-      $json_table = [
-        '#type' => 'table',
-        '#header' => [t('Unused JSON keys')],
-        '#rows' => $unused_rows,
-        '#empty' => t('No content has been found.'),
-      ];
+      $json_table = [];
+      $var_table = [];
 
-      // Try to Ensure we're using the twig from user's input instead of the entity's
-      // default.
       try {
         $input = $form_state->getUserInput();
-        $entity->set('twig', $input['twig'][0], FALSE);
+        $preview_template = $input['twig'];
+        $entity->set('twig', $preview_template[0], FALSE);
         $render = $entity->renderNative($context);
+        $used_vars = $entity->getTwigVariablesUsed();
+        $flattened_keys = MetadataDisplayForm::flattenKeys($jsondata);
+        $data_json = $flattened_keys;
+        ksort($data_json);
+        $used_keys = [];
+        $used_vars_other = [];
+        foreach($used_vars as $used_var) {
+          $used_var_path = $used_var['path'];
+          $used_var_line = $used_var['line'];
+          $used_var_parent_path = isset($used_var['parent_path']) ? $used_var['parent_path'] : '';
+          if (str_starts_with($used_var_path, 'data.')) {
+            $used_var_exploded = explode('.', $used_var_path);
+            array_push($used_keys, $used_var_path);
+            if (array_key_exists($used_var_path,$data_json)) {
+              $data_json[$used_var_path]['used'] = 'Used';
+              $data_json[$used_var_path]['line'] = $used_var_line;
+            }
+            if (count($used_var_exploded) > 2) {
+              $used_var_parts = array_slice($used_var_exploded,0, 2);
+              $used_var_part = implode('.', $used_var_parts);
+              if (array_key_exists($used_var_part,$data_json)) {
+                $data_json[$used_var_part]['used'] = 'Used';
+                $data_json[$used_var_part]['line'] = $used_var_line;
+
+              }
+
+            }
+          }
+          else {
+            if (!empty($used_var_parent_path) && str_starts_with($used_var_parent_path, 'data.') && array_key_exists($used_var_parent_path, $data_json)) {
```

Comment:
Suggestion . In general doing `??` for array key existence is faster. Also  isset($used_var_parent_path[$data_json])  is faster

Reply:
Resolved via edad28026d576acb959a73f15fdca4c20d7110e4.